external/fv3fit/fv3fit/reservoir/validation.py

```diff
@@ -1,9 +1,147 @@
 import numpy as np
 from scipy.ndimage import generic_filter
-from typing import Union, Optional, Sequence
+from typing import Union, Optional, Sequence, Hashable, Mapping
 import xarray as xr
 import tensorflow as tf
 import wandb
+import matplotlib.pyplot as plt
+from toolz import curry
+from io import BytesIO
+from PIL import Image
+import logging
+
+from fv3fit.reservoir.adapters import ReservoirAdapterType
+
+
+UNITS = {
+    "sst": "K",
+}
+
+logger = logging.getLogger(__name__)
+
+
+def _run_one_step_predictions(synced_model, inputs, hybrid):
+
+    # run one steps
+    predictions = []
+    for i in range(len(inputs.time)):
+        synced_model.increment_state(inputs.isel(time=i))
+        current_hybrid = hybrid.isel(time=i) if hybrid else {}
+        predictions.append(synced_model.predict(current_hybrid))
+
+    return predictions
+
+
+def _get_slice(src_len, dst_len):
+    """
+    src_len: length of previous state dimension to be inserted into current state
+    dst_len: length of current state dimension
+    """
+    if src_len == dst_len:
+        sl = slice(None)
+    elif src_len < dst_len:
+        diff = dst_len - src_len
+        overlap = diff // 2
+        sl = slice(overlap, -overlap)
+    else:
+        raise ValueError("src_len must be <= dst_len")
+
+    return sl
+
+
+def _insert_tile_to_overlap(current: xr.DataArray, previous: xr.DataArray):
+    # we can't grab the halos for offline rollouts because there is no prediction
+    # for other tiles.  Instead just grab the original overlap, which is *very*
+    # optimistic for forecasting...
```

Comment:
Is this a large enough concern that we should just return constant values for rollout error if overlap>0?

Reply:
I don't think it's a super large concern.  I usually am not comparing inter-experiment offline skill to a fine degree since the spread on similar hyperparameters is pretty wide.  I just check that the rollouts aren't blowing up or in the 75th percentile of RMSE.